### PR TITLE
Guide Availability performance enhancements

### PR DESCRIPTION
**Pull request by toddburnside**
A couple of things have changed:

- Based on a suggestion by Shane, it is using a stream version of `AGS.agsAnalysis` to analyze the candidates and stops if/when we have the data we need. In some example observations I checked, this resulted in reduction of almost an order of magnitude, while in others it had little effect.
- Only analyze the angles required based on the PosAngleConstraint.

### CODE REVIEW

**Review by swalker2m — APPROVED** [review]
This looks like a good improvement.  Over in `callGaia` we materialize the stream of candidates into a `List` in memory.  I wonder if we could stream it all the way through and not read them completely unless necessary?  Anyway :+1:

File: modules/service/src/main/scala/lucuma/odb/service/GuideService.scala [hunk 1]
```diff
@@ -285,14 +289,31 @@ object GuideService {
...
+            // These should always be big values (until nonsidereal), but if it is 0, we'd go infinite loop.
+            val daysTilInvalid = scala.math.max((invalidThreshold / speed * 365).intValue, 1)
+            Timestamp.fromInstantTruncated(startDate.toInstant.plus(daysTilInvalid, ChronoUnit.DAYS))
+              .getOrElse(Timestamp.Max)
+          }
+        }
+
+      extension (timestamp: Timestamp)
+        def laterOf(time2: Timestamp): Timestamp =
+          if (timestamp > time2) timestamp else time2
```
[the remark this answers is not in the document]
Comment (toddburnside) [reply modules/service/src/main/scala/lucuma/odb/service/GuideService.scala:314]: Oh, thanks.

File: build.sbt [hunk 2]
```diff
@@ -54,6 +54,7 @@ lazy val schema =
     .in(file("modules/schema"))
     .settings(
       name := "lucuma-odb-schema",
+      libraryDependencySchemes += "org.typelevel" %% "cats-parse" % "always", // Workaround until http4s updates to 1.0.0 as well
```
[the remark this answers is not in the document]
Comment (toddburnside) [reply build.sbt:57]: Apparently, it says "don't worry about the dependency conflict, just always use the lastest one". Arman showed me this from grackle. Once http4s is on cats-parse 1.0.0 I'll remove it.

**Comment by swalker2m** [comment]
> I would like to, but the problem is that we may, or may not, need to process the candidates more than once for different time periods.

Ah, ok.  Out of curiosity, what is the scenario in which they have to be processed multiple times for different time periods? In the same call?

**Comment by toddburnside** [comment]
> Ah, ok. Out of curiosity, what is the scenario in which they have to be processed multiple times for different time periods? In the same call?

If any of the science targets or candidates move more than 100mas from the start of a time period, we have to end the time period and start a new time period. This means redoing the ags analysis for the candidates for the new time and recalculating the available angles. If the target happens to be Barnard's Star, we have to do that many times in a sixth month period. (Barnard's Star moves 10 arcsec/year)

**Comment by toddburnside** [comment]
> I see, that makes sense. Could it be flipped to analyze the target with each candidate successively over the entire time period? That is for candidate _n_, find the angles and time periods that are valid for the target and _n_ for the entire 6 months. Maybe that doesn't make sense or is prohibitively difficult. I only ask thinking that for crowded fields keeping all the candidates in memory could be an issue maybe, beyond the cost to parse them all etc. idk. At any rate, I don't want to hold up progress now!

That would definitely be more complicated. It might be doable, but I'd have to think through it. It would require tracking a lot more data during the processing. I'm not sure how timeouts for reading streams over http work and if we'd run into a problem there. Maybe before trying to dig into all of that, I'll ask Andy for a star in a very dense star field and do a query with `widestConstraints` to see how many candidates we're talking about. If it is a very large number, and timeouts are an issue we could do something like read into a temporary table?

**Pull request merged**